### PR TITLE
Supports default ConfigMap params

### LINKED ISSUE #7: Need to support maintaining default ConfigMap params
Current autoscaler only takes the responsibility to fetch ConfigMap from the apiserver. When there is no ConfigMap created or the previous ConfigMap on apiserver got accidentally deleted, autoscaler will fail. It would be good that this autoscaler also provides the ability to create a default ConfigMap when not present, and the default ConfigMap params could(or should) be provided by user.

Here is one specific use case of autoscaler with kube-dns addon of kubernetes:
The plan to use this autoscaler with kube-dns is to make it also an addon and put its manifest into the /addon folder. By doing so, the addon manager will manage its creation and update. A problem will happen to the ConfigMap if we also treat it as an addon resource. Addon manager assumes ownership over all supplied fields in the initial config, which means the user would not be able to modify the autoscale params(ConfigMap) if they don't have the permission to modify the manifests locate on the master node. Without the ability to configure the scaling params, this autoscaler is much less useful.

An obvious way to solve above problem is to only create the ConfigMap once on startup instead of handing over it to the addon manager. But that leads to another issue that what if the ConfigMap got accidentally deleted by user? The autoscaler will be functionless if no one re-create the ConfigMap params.

To make the autoscaling feature both flexible and robust, a quick proposal below:
- Add one more flag (like --default-params) for user to provide the default params for scaling. The autoscaler will create/update the corresponding ConfigMap on startup if flag presents.
- The working logic remains the same, user could still modify the ConfigMap params on the fly. But if fetching ConfigMap from apiserver fails, autoscaler should re-create the ConfigMap with the given default params.
- If default params not provided, autoscaler acts as before and will not manage the ConfigMap.

@bowei @thockin @bprashanth

### PR DESCRIPTION
**Pull request by MrHohn**
Fix #7.

Added `--default-params` flag for passing in the default params(in JSON format) of ConfigMap. If corresponding ConfigMap not exist / deleted on apiserver, autoscaler will create it with the default params, given name and namespace. It is needed for properly integrating this autoscaler with kube-dns addon on kubernetes.

README is updated and examples are added.

@bowei @thockin @bprashanth

### CODE REVIEW

**Comment by MrHohn** [comment]
Rebased.

File: cmd/cluster-proportional-autoscaler/options/options.go [hunk 1]
```diff
@@ -90,7 +121,8 @@ func (c *AutoScalerConfig) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&c.Target, "target", c.Target, "Target to scale. In format: deployment/*, replicationcontroller/* or replicaset/* (not case sensitive).")
 	fs.StringVar(&c.ConfigMap, "configmap", c.ConfigMap, "ConfigMap containing our scaling parameters.")
 	fs.StringVar(&c.Namespace, "namespace", c.Namespace, "Namespace for all operations, fallback to the namespace of this autoscaler(through MY_POD_NAMESPACE env) if not specified.")
-	fs.StringVar(&c.Mode, "mode", c.Mode, "Control mode(linear/ladder). Default is the linear mode.")
+	fs.StringVar(&c.Mode, "mode", c.Mode, "Control mode(linear/ladder).")
 	fs.IntVar(&c.PollPeriodSeconds, "poll-period-seconds", c.PollPeriodSeconds, "The time, in seconds, to check cluster status and perform autoscale.")
 	fs.BoolVar(&c.PrintVer, "version", c.PrintVer, "Print the version and exit.")
+	fs.String("default-params", "", "Default parameters(JSON format) for auto-scaling. Will create/re-create a ConfigMap with this default params if not present.")
```
Comment (bowei) [inline cmd/cluster-proportional-autoscaler/options/options.go]: if ConfigMap is not present (may be confusing with if --default-params is not present)

File: pkg/autoscaler/autoscaler_server.go [hunk 2]
```diff
@@ -91,16 +93,16 @@ func (s *AutoScaler) pollAPIServer() {
 	// Query the apiserver for the cluster status --- number of nodes and cores
 	clusterStatus, err := s.k8sClient.GetClusterStatus()
 	if err != nil {
-		glog.Errorf("error while counting nodes: %v\n", err)
+		glog.Errorf("Error while getting cluster status: %v\n", err)
 		return
 	}
-	glog.V(4).Infof("Total nodes %5d, schedulable nodes: %5d\n", clusterStatus.TotalNodes, clusterStatus.SchedulableNodes)
-	glog.V(4).Infof("Total cores %5d, schedulable cores: %5d\n", clusterStatus.TotalCores, clusterStatus.SchedulableCores)
+	glog.V(6).Infof("Total nodes %5d, schedulable nodes: %5d\n", clusterStatus.TotalNodes, clusterStatus.SchedulableNodes)
```
Comment (bowei) [inline pkg/autoscaler/autoscaler_server.go]: why bump to v(6)? v(4) is already not enabled by default

File: pkg/autoscaler/controller/controller.go [hunk 3]
```diff
@@ -22,6 +22,6 @@ import (
 
 type Interface interface {
 	GetExpectedReplicas(k8sclient.ClusterStatus) (int32, error)
-	SyncConfig(k8sclient.ConfigMap) error
+	SyncConfig(*k8sclient.ConfigMap) error
```
Comment (bowei) [inline pkg/autoscaler/controller/controller.go]: document? also, should be called Controller, not Interface?

File: pkg/autoscaler/k8sclient/k8sclient.go [hunk 4]
```diff
@@ -31,7 +31,9 @@ import (
 
 // Interface - Wraps all needed client functionalities for autoscaler
 type Interface interface {
```
Comment (bowei) [inline pkg/autoscaler/k8sclient/k8sclient.go]: should be called ConfigClient?

File: pkg/autoscaler/k8sclient/k8sclient.go [hunk 5]
```diff
@@ -31,7 +31,9 @@ import (
...
-	FetchConfigMap(namespace, configmap string) (ConfigMap, error)
+	FetchConfigMap(namespace, configmap string) (*ConfigMap, error)
```
Comment (bowei) [inline pkg/autoscaler/k8sclient/k8sclient.go]: documentation?

File: pkg/autoscaler/k8sclient/k8sclient.go (hunk 4 above)
Comment (MrHohn) [inline pkg/autoscaler/k8sclient/k8sclient.go]: I think it is doing more than ConfigClient. What about `K8sClient`?

File: cmd/cluster-proportional-autoscaler/options/options.go (hunk 1 above)
Comment (MrHohn) [inline cmd/cluster-proportional-autoscaler/options/options.go]: Fixed.

File: pkg/autoscaler/autoscaler_server.go (hunk 2 above)
Comment (MrHohn) [inline pkg/autoscaler/autoscaler_server.go]: I forgot the reason, but V(4) seems reasonable now. Changed back.

File: pkg/autoscaler/controller/controller.go (hunk 3 above)
Comment (MrHohn) [inline pkg/autoscaler/controller/controller.go]: Changed to Controller and added comments.

**Comment by MrHohn** [comment]
Thanks for reviewing this! Repushed commits.

**Comment by bowei** [comment]
````
package main

import (
	"fmt"
	"pflag"
)

type MyStuff map[string]string

func (ms *MyStuff) Set(raw string) error {
	return nil
}

func (ms *MyStuff) String() string {
	return "blah"
}

func (ms *MyStuff) Type() string {
	return "string"
}

func main() {
	fmt.Println("hello")

	var ms MyStuff
	pflag.Var(&ms, "ms", "blah")
}
````

File: cmd/cluster-proportional-autoscaler/options/options.go [hunk 6]
```diff
@@ -63,6 +65,19 @@ func (c *AutoScalerConfig) ValidateFlags() error {
 		errorsFound = true
 		glog.Errorf("--poll-period-seconds cannot be less than 1\n")
 	}
+	params, err := pflag.CommandLine.GetString("default-params")
```
Comment (MrHohn) [inline cmd/cluster-proportional-autoscaler/options/options.go]: Fixed.

**Pull request merged**